File: fushi/test/reader/reader_dblclick_clear_selection_behavior_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

/// TODO-1028 / BUG-481: a native double-click that establishes a text selection
/// must NOT hijack single-tap lookup, and must NOT block the furigana whole-page
/// double-click toggle.
///
/// 回归背景：WebView 原生双击在正文上建立蓝色框选（`selectstart` 抑制只挡拖动起手
/// 400ms 内、不挡双击；正文 `_fushiReaderMouseNativeTextStart=true` 放过双击选词），
/// 盖住单击查词的 CSS Highlight，并把振假名整页切换（`_buildFuriganaJs` 'toggle' 分支
/// 的 dblclick handler 带 `if (sel && !sel.isCollapsed) return` 守卫）绊住——双击自己
/// 产生的选区让守卫早退，振假名不切换。修复在 `_buildReaderSetupScript` 加 capture 阶段
/// dblclick → `removeAllRanges()`，先于振假名 bubble handler 清掉选区，振假名切换反而
/// 恢复正常。
///
/// 这里在 `flutter test` 内通过 Node 真执行从 webview.part.dart 抽取的两个真实
/// dblclick handler（伪 DOM + capture→bubble 顺序派发），断言双击后选区被清
/// （isCollapsed===true）且 show-all-rt 仍被 toggle（振假名保留）。撤掉修复，Node
/// 断言失败、本 Dart 守卫转红。
///
/// 当本机/CI 没有 node 时自动 skip，但本地与装有 node 的环境都会真跑，提供静态守卫
/// 缺失的行为级覆盖。
void main() {
  test(
      'double-click clears native selection but still toggles furigana '
      '(executes reader handlers via node)', () async {
    final String? nodeExe = _resolveNode();
    if (nodeExe == null) {
      markTestSkipped('node not found on PATH; skipping JS behavior execution');
      return;
    }

    final File jsTest = File(
      'test/reader/reader_dblclick_clear_selection_behavior_test.js',
    );
    expect(jsTest.existsSync(), isTrue,
        reason: 'behavior harness ${jsTest.path} must exist');

    final ProcessResult result = await Process.run(
      nodeExe,
      <String>[jsTest.path],
      workingDirectory: Directory.current.path,
    );

    expect(
      result.exitCode,
      0,
      reason: 'reader dblclick handler JS behavior test failed.\n'
          'stdout:\n${result.stdout}\nstderr:\n${result.stderr}',
    );
    expect(
      result.stdout.toString(),
      contains('all assertions passed'),
      reason: 'behavior harness must reach its success marker',
    );
  });
}

/// Resolve a usable `node` executable, returning null when none is on PATH.
String? _resolveNode() {
  final List<String> candidates =
      Platform.isWindows ? <String>['node.exe', 'node'] : <String>['node'];
  for (final String name in candidates) {
    try {
      final ProcessResult probe = Process.runSync(name, <String>['--version']);
      if (probe.exitCode == 0) {
        return name;
      }
    } on ProcessException {
      // Not found; try next candidate.
    }
  }
  return null;
}

File: fushi/test/reader/reader_dblclick_clear_selection_behavior_test.js
// TODO-1028 / BUG-481 behavior test: a native double-click that establishes a
// text selection must NOT hijack lookup or block the furigana whole-page toggle.
//
// Root cause: the WebView native double-click builds a (blue) text selection that
// covers the single-tap lookup CSS Highlight and trips the furigana dblclick
// toggle handler, whose guard `if (sel && !sel.isCollapsed) return` bails out when
// it sees the selection the double-click itself just produced. The fix adds a
// CAPTURE-phase dblclick listener that removeAllRanges() the native selection
// BEFORE the (bubble-phase) furigana handler runs, so the furigana toggle sees a
// collapsed selection and fires normally.
//
// This test EXECUTES both real handlers, extracted verbatim from
// webview.part.dart: (1) the new capture-phase clear-selection dblclick listener
// inside _buildReaderSetupScript, and (2) the 'toggle' furigana dblclick handler
// from _buildFuriganaJs. It dispatches a dblclick capture-then-bubble (browser
// order) over a non-collapsed selection and asserts the selection is cleared
// (isCollapsed === true) AND show-all-rt was still toggled (furigana preserved).
// Reverting the fix (dropping removeAllRanges) leaves the selection non-collapsed,
// so the furigana guard bails and show-all-rt is NOT toggled -> red.
//
// Run: node fushi/test/reader/reader_dblclick_clear_selection_behavior_test.js
// (also driven from reader_dblclick_clear_selection_behavior_test.dart so it runs
//  inside `flutter test`).

const assert = require('assert');
const fs = require('fs');
const path = require('path');
const vm = require('vm');

const readerPath = path.resolve(
  __dirname,
  '../../lib/src/pages/implementations/reader_fushi/webview.part.dart',
);
const source = fs.readFileSync(readerPath, 'utf8');

// (1) Extract the capture-phase clear-selection dblclick listener from
// _buildReaderSetupScript. Anchor on the TODO-1028 comment and capture the whole
// addEventListener('dblclick', ..., true) statement.
const captureMatch = source.match(
  /document\.addEventListener\('dblclick',\s*function\(\)\s*\{\s*var sel = window\.getSelection && window\.getSelection\(\);\s*if \(sel && !sel\.isCollapsed\) sel\.removeAllRanges\(\);\s*\},\s*true\);/,
);
assert.ok(captureMatch, 'missing TODO-1028 capture-phase clear-selection dblclick listener');
const captureListener = captureMatch[0];

// (2) Extract the furigana 'toggle' dblclick handler from _buildFuriganaJs. It is
// inside a Dart triple-quoted string; grab the addEventListener statement only.
const toggleMatch = source.match(
  /document\.addEventListener\('dblclick',\s*function\(\)\s*\{\s*var sel = window\.getSelection\(\);\s*if \(sel && !sel\.isCollapsed\) return;\s*document\.body\.classList\.toggle\('show-all-rt'\);\s*\}\);/,
);
assert.ok(toggleMatch, "missing _buildFuriganaJs 'toggle' dblclick handler");
const furiganaListener = toggleMatch[0];

function makeHarness() {
  // capture-phase listeners run before bubble-phase ones (browser order). The
  // fake document records (listener, useCapture) and dispatch() replays capture
  // listeners first, then bubble.
  const captureListeners = [];
  const bubbleListeners = [];

  // A live, mutable selection model: starts non-collapsed (the double-click just
  // selected a word); removeAllRanges() collapses it.
  const selection = {
    isCollapsed: false,
    removeAllRanges() { this.isCollapsed = true; },
  };

  const bodyClassList = {
    _set: new Set(),
    toggle(name) {
      if (this._set.has(name)) { this._set.delete(name); } else { this._set.add(name); }
    },
    has(name) { return this._set.has(name); },
  };

  const documentObj = {
    body: { classList: bodyClassList },
    addEventListener(name, fn, useCapture) {
      if (name !== 'dblclick') { return; }
      (useCapture ? captureListeners : bubbleListeners).push(fn);
    },
  };

  const windowObj = {
    getSelection() { return selection; },
  };

  const sandbox = {
    document: documentObj,
    window: windowObj,
  };

  // Register BOTH handlers in source order (capture listener appears first in the
  // setup script; furigana toggle is appended via _buildFuriganaJs). Registration
  // order is irrelevant — dispatch order is governed by useCapture.
  const prepared = '(function(){\n' + captureListener + '\n' + furiganaListener + '\n})();';
  vm.createContext(sandbox);
  vm.runInContext(prepared, sandbox, { filename: 'reader-dblclick-handlers.js' });

  function dispatchDblclick() {
    // Browser phase order: capture listeners first, then bubble listeners.
    for (const fn of captureListeners) { fn(); }
    for (const fn of bubbleListeners) { fn(); }
  }

  return { dispatchDblclick, selection, bodyClassList };
}

// Test 1: a double-click with a live (non-collapsed) selection -> the capture
// listener clears it, and because it ran first the furigana toggle still fires.
(function () {
  const h = makeHarness();
  assert.strictEqual(h.selection.isCollapsed, false, 'precondition: selection starts non-collapsed (the dblclick selected a word)');
  h.dispatchDblclick();
  assert.strictEqual(
    h.selection.isCollapsed, true,
    'native double-click selection must be cleared (removeAllRanges) so it stops hijacking lookup',
  );
  assert.strictEqual(
    h.bodyClassList.has('show-all-rt'), true,
    'furigana whole-page toggle must still fire on double-click (the capture clear runs first so the !isCollapsed guard no longer bails)',
  );
})();

// Test 2: a second double-click toggles show-all-rt back off (proves it is a real
// toggle, not a one-way set), and the now-collapsed selection stays collapsed.
(function () {
  const h = makeHarness();
  h.dispatchDblclick(); // on -> show-all-rt added
  // Simulate the browser re-establishing a selection on the next double-click.
  h.selection.isCollapsed = false;
  h.dispatchDblclick(); // off -> show-all-rt removed
  assert.strictEqual(h.selection.isCollapsed, true, 'second double-click selection must also be cleared');
  assert.strictEqual(
    h.bodyClassList.has('show-all-rt'), false,
    'second double-click must toggle show-all-rt back off',
  );
})();

console.log('reader_dblclick_clear_selection_behavior_test.js: all assertions passed');
